src/services/ipfs/backend.rs

```diff
@@ -0,0 +1,209 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ipfs::Backend")
+            .field("root", &self.root)
+            .finish()
+    }
+}
+
+impl Backend {
+    /// Create a default builder for ipfs.
+    pub fn build() -> Builder {
+        Builder::default()
+    }
+
+    pub(crate) fn new(root: String, client: IpfsClient) -> Self {
+        Self { root, client }
+    }
+
+    pub(crate) fn from_iter(
+        it: impl Iterator<Item = (String, String)>,
+    ) -> io::Result<Self> {
+        let mut builder = Builder::default();
+
+        for (key, val) in it {
+            let val = val.as_str();
+            match key.as_ref() {
+                "root" => builder.root(val),
+                "endpoint" => builder.endpoint(val),
+                _ => continue,
+            };
+        }
+
+        builder.build()
+    }
+
+    pub(crate) fn get_abs_path(&self, path: &str) -> String {
+        if path == self.root {
+            return path.to_string();
+        }
+
+        format!("{}{}", self.root, path.trim_start_matches(&self.root))
+    }
+
+    pub(crate) async fn files_stat(&self, path: &str) -> io::Result<ObjectMetadata> {
+        let mut meta = ObjectMetadata::default();
+
+        let res = self
+            .client
+            .files_stat(path)
+            .await
+            .map_err(crate::error::other)?;
+
+        let mode: ObjectMode = match res.typ.as_str() {
+            "file" => ObjectMode::FILE,
+            "directory" => ObjectMode::DIR,
+            _ => ObjectMode::Unknown,
+        };
+
+        meta.set_mode(mode).set_content_length(res.size);
+
+        Ok(meta)
+    }
+
+    pub(crate) async fn files_create(&self, path: &str) -> io::Result<()> {
+        self.client
+            .files_write(path, true, false, io::empty())
+            .await
+            .map_err(crate::error::other)
+    }
+
+    pub(crate) async fn files_read(
+        &self,
+        path: &str,
+        offset: Option<i64>,
+        count: Option<i64>,
+    ) -> io::Result<BytesReader> {
+        let req = ipfs_api_backend_hyper::request::FilesRead {
+            path,
+            offset,
+            count,
+        };
+        let reader = self
+            .client
+            .files_read_with_options(req)
+            .map_err(crate::error::other)
+            .into_async_read();
+        Ok(Box::new(reader))
+    }
+
+    pub(crate) async fn files_delete(&self, path: &str) -> io::Result<()> {
+        self.client
+            .files_rm(path, false)
+            .await
+            .map_err(crate::error::other)
+    }
+
+    pub(crate) async fn files_list(&self, path: &str) -> io::Result<FilesLsResponse> {
+        self.client
+            .files_ls(Some(path))
+            .await
+            .map_err(crate::error::other)
+    }
+
+    pub(crate) async fn files_write(&self, path: &str, data: Bytes) -> io::Result<()> {
+        self.client
+            .files_write(path, true, true, data.reader())
```

Comment:
Sadly, we can't implement `Read` for `Body`. Maybe we need to build a HTTP request directly like we do for `s3` service.

FYI: https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-files-write

BTW: most API calls is so simple that we don't need to use `ipfs-api` SDK.